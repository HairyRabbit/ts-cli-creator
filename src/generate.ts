import { ts, printNode } from 'ts-morph'
import * as prettier from 'prettier'

export function generateCallableChain(calls: ts.CallExpression[], expr: ts.Expression): ts.CallExpression {
  return calls.reverse().reduce((acc: any, call: any) => {
    return (expr: ts.Expression) => replaceCallableProperty(acc(call), expr)
  }, (a: any) => a)(expr)
}

function replaceCallableProperty(call: ts.CallExpression, expr: ts.Expression): ts.CallExpression {
  return ts.createCall(
    ts.createPropertyAccess(
      expr,
      call.expression as ts.Identifier
    ),
    call.typeArguments,
    call.arguments
  )
}

const DEFAULT_NAME: string = `yargs`

export interface GenerateOptions {
  optionCalls: ts.CallExpression[]
  name: string
  strict: boolean,
  help: boolean
}

const DEFAULT_GENERATE_OPTIONS: GenerateOptions = {
  optionCalls: [],
  name: DEFAULT_NAME,
  strict: true,
  help: true
}

export default function generate(options: Partial<GenerateOptions> = {}): ts.Node[] {
  const { optionCalls, name, strict, help } = { ...DEFAULT_GENERATE_OPTIONS, ...options }
  
  const acc: ts.CallExpression[] = []

  if(strict) {
    acc.push(ts.createCall(ts.createIdentifier('strict'), undefined, []))
  }
  
  if(0 !== optionCalls.length) {
    optionCalls.forEach(call => acc.push(call))
  }
  
  if(help) {
    acc.push(ts.createCall(ts.createIdentifier('help'), undefined, []))
    acc.push(ts.createCall(ts.createIdentifier('alias'), undefined, [
      ts.createStringLiteral('help'),
      ts.createStringLiteral('h')
    ]))
  }

  const callableChainNodes = generateCallableChain(acc, ts.createIdentifier(name))
  
  const yargsNode =
  ts.createVariableStatement(
    undefined,
    ts.createVariableDeclarationList(
      [
        ts.createVariableDeclaration(
          ts.createIdentifier('args'),
          ts.createTypeReferenceNode(
            ts.createQualifiedName(
              ts.createIdentifier(`yargs`),
              ts.createIdentifier(`Arguments`)
            ),
            [
              ts.createTypeReferenceNode(ts.createIdentifier('Options'), undefined)
            ]
          ),
          ts.createPropertyAccess(
            callableChainNodes,
            ts.createIdentifier(`argv`)
          )
        )
      ], ts.NodeFlags.Const
    )
  )

  const deconstructNode =
  ts.createVariableStatement(
    undefined,
    ts.createVariableDeclarationList(
      [
        ts.createVariableDeclaration(
          ts.createObjectBindingPattern(
            [
              ts.createBindingElement(
                undefined,
                undefined,
                ts.createIdentifier('_'),
                undefined
              ),
              ts.createBindingElement(
                undefined,
                undefined,
                ts.createIdentifier('$0'),
                undefined
              ),
              ts.createBindingElement(
                ts.createToken(ts.SyntaxKind.DotDotDotToken),
                undefined,
                ts.createIdentifier('options'),
                undefined
              )
            ]
          ),
          undefined,
          ts.createIdentifier('args')
        )
      ],
      ts.NodeFlags.Const
    )
  )

  const applyCommandNode =
  ts.createExpressionStatement(
    ts.createCall(ts.createIdentifier('command'), undefined, [
      ts.createIdentifier('options')
    ])
  )

  return generateWrapper([ yargsNode, deconstructNode, applyCommandNode ])

  // return yargsNode
}

export function render(nodes: ts.Node[]) {
  const code = nodes.map(node => printNode(node)).join(`\n`)
  return prettier.format(code, { parser: 'typescript' })
}

export function generateWrapper(body: ts.Statement[]): ts.Node[] {
  return [
    ts.createImportDeclaration(
      undefined,
      undefined,
      ts.createImportClause(
        undefined,
        ts.createNamespaceImport(
          ts.createIdentifier('yargs')
        ),
      ),
      ts.createStringLiteral('yargs')
    ),
    
    ts.createImportDeclaration(
      undefined,
      undefined,
      ts.createImportClause(
        ts.createIdentifier('command'), 
        ts.createNamedImports(
          [
            ts.createImportSpecifier(
              undefined,
              ts.createIdentifier(`Options`)
            )
          ]
        )
      ),
      ts.createStringLiteral('./index')
    ),

    ts.createFunctionDeclaration(
      undefined,
      [
        ts.createModifier(ts.SyntaxKind.ExportKeyword),
        ts.createModifier(ts.SyntaxKind.DefaultKeyword),
      ],
      undefined,
      ts.createIdentifier('main'),
      undefined,
      [],
      ts.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword),
      ts.createBlock(body, true)
    )
  ]
}
