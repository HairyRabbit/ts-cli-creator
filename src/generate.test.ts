import * as ts from 'typescript'
import generate, { render, generateWrapper, GenerateOptions } from './generate'

describe(`generate()`, () => {
  test(`default`, () => {
    const options: Partial<GenerateOptions> = {
      optionCalls: [ ts.createCall(ts.createIdentifier('foo'), [], undefined) ]
    }
    const result = render(generate(options))
    console.log(result)
  })
})

describe(`generateWrapper()`, () => {
  test(`basic`, () => {
    const result: string = render(generateWrapper(ts.createLiteral(42) as any))
    expect(result).toMatchSnapshot()
  })
})
