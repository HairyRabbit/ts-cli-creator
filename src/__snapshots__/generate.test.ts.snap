// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateWrapper() basic 1`] = `
"import * as yargs from \\"yargs\\";
import command, { Options } from \\"./index\\";
export default function main(): void {}
"
`;
